core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -346,19 +346,29 @@ private[deploy] class Master(
           logInfo("Worker has been re-registered: " + workerId)
           worker.state = WorkerState.ALIVE
 
-          val validExecutors = executors.filter(exec => idToApp.get(exec.appId).isDefined)
+          val (validExecutors, invalidExecutors) =
+            executors.partition(exec => idToApp.get(exec.appId).isDefined)
           for (exec <- validExecutors) {
```

Comment:
`validExecutors` may contains executors launched after master sends `RegisteredWorker`. Then `app.addExecutor` and `worker.addExecutor` will be called twice. So I made them idempotent.